server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -0,0 +1,1192 @@
+         .emit();
+      return new NoopQueryRunner<>();
+    }
+
+    final QueryToolChest<T, Query<T>> toolchest = factory.getToolchest();
+    final Function<Query<T>, ServiceMetricEvent.Builder> builderFn =
+        new Function<Query<T>, ServiceMetricEvent.Builder>()
+        {
+
+          @Override
+          public ServiceMetricEvent.Builder apply(@Nullable Query<T> input)
+          {
+            return toolchest.makeMetricBuilder(query);
+          }
+        };
+    final boolean skipIncrementalSegment = query.getContextValue(CONTEXT_SKIP_INCREMENTAL_SEGMENT, false);
+
+    return toolchest.mergeResults(
+        factory.mergeRunners(
+            queryExecutorService,
+            FunctionalIterable
+                .create(specs)
+                .transform(
+                    new Function<SegmentDescriptor, QueryRunner<T>>()
+                    {
+                      @Override
+                      public QueryRunner<T> apply(final SegmentDescriptor descriptor)
+                      {
+                        final PartitionHolder<Sink> holder = sinkTimeline.findEntry(
+                            descriptor.getInterval(),
+                            descriptor.getVersion()
+                        );
+                        if (holder == null) {
+                          return new ReportTimelineMissingSegmentQueryRunner<>(descriptor);
+                        }
+
+                        final PartitionChunk<Sink> chunk = holder.getChunk(descriptor.getPartitionNumber());
+                        if (chunk == null) {
+                          return new ReportTimelineMissingSegmentQueryRunner<>(descriptor);
+                        }
+
+                        final Sink theSink = chunk.getObject();
+
+                        return new SpecificSegmentQueryRunner<>(
+                            new MetricsEmittingQueryRunner<>(
+                                emitter,
+                                builderFn,
+                                new BySegmentQueryRunner<T>(
+                                    theSink.getSegment().getIdentifier(),
+                                    descriptor.getInterval().getStart(),
+                                    factory.mergeRunners(
+                                        MoreExecutors.sameThreadExecutor(),
+                                        Iterables.transform(
+                                            theSink,
+                                            new Function<FireHydrant, QueryRunner<T>>()
+                                            {
+                                              @Override
+                                              public QueryRunner<T> apply(final FireHydrant hydrant)
+                                              {
+                                                if (skipIncrementalSegment && !hydrant.hasSwapped()) {
+                                                  return new NoopQueryRunner<>();
```

Comment:
what is the end experience for hte user when this is returned? Will they understand what is happening?